src/clj_irods/icat.clj

```diff
@@ -0,0 +1,207 @@
+
+(defn filtered-cached-listings
+  [irods user zone path & {:keys [entity-type sort-column sort-direction limit offset info-types]}]
+  (let [all (all-cached-listings irods user zone path)]
+    (when all
+      (let [filters [entity-type info-types sort-column sort-direction limit offset]
+            cached  (filter-listings* filters @all)]
+        (when (seq cached)
+          (delay cached))))))
+
+(defn- keys-in [m]
+  (if (map? m)
+    (vec
+      (mapcat (fn [[k v]]
+                (let [sub (keys-in v)
+                      nested (map #(into [k] %) (filter (comp not empty?) sub))]
+                  (if (seq nested)
+                    nested
+                    [[k]])))
+              m))
+    []))
+
+;; NOTE: this merges only directly adjacent ranges -- overlapping ranges will remain separate. This could be less than ideal.
+;; e.g. if we have three ranges, (0 30) (10 40) (20 50), none of them will merge even though we have everything in (0 50)
+;; I'm hoping this doesn't matter much in practice, since it's usually not all that useful to grab overlapping ranges.
+(defn- merge-listing
+  [extant to-merge limit offset]
+  (let [ranges (into {} (mapcat (fn [[k v]] (mapv #(vector (+ % k) [k %]) (keys v))) extant))]
+    (if (contains? ranges offset) ;; is there an existing range that ends at our start point
+      (let [[old-limit old-offset] (get ranges offset) ;; there is, so extend it with these new values
+            extant-values (get-in extant [old-limit old-offset])
+            new-values    (concat extant-values to-merge)]
+        (into {} (remove (fn [[k v]] (empty? v))
+          (-> extant
+              (update-in [old-limit] dissoc old-offset) ;; effectively, dissoc-in [old-limit old-offset]
+              (assoc-in [(if (nil? limit) nil (+ old-limit limit)) old-offset] new-values)))))
+      (assoc-in extant [limit offset] to-merge)))) ;; there is not, so just stick the merge in
+
+(defn merge-listings
+  [listings]
+  (let [listings (if (delay? listings) @listings listings)
+        key-paths (map (partial take 6) (keys-in listings))]
+    (reduce
+      (fn [m ks]
+        (let [ks4            (take 4 ks)
+              [limit offset] (drop 4 ks)]
+          (if-let [extant (get-in m ks4)]
+            (assoc-in m ks4 (merge-listing extant (get-in listings ks) limit offset))
+            (assoc-in m ks (get-in listings ks)))))
+      {}
+      key-paths)))
+
+(defn get-range
+  "Given merged listings selected down to limit-offset section, get a section of results for a limit & offset"
+  [merged-listings limit offset]
+  (let [all-ranges (mapv (partial take 2) (keys-in merged-listings))
+        unbounded-limit-offset (filter (fn [[l o]] (and (nil? l) (>= offset o))) all-ranges)
+        bounded-limit-offset (and
+                               (not (nil? limit))
+                               (filter (fn [[l o]] (and (not (nil? l)) (<= limit l) (>= offset o))) all-ranges))
```

Comment:
i.e. which routes into `merged-listings` have a non-nil limit, but a limit higher than ours and an offset lower

as I write this, I think this may need amendment because the selected limit needs to be greater than or equal to the requested limit _plus_ the difference in offsets. i.e., given (0 30), this would say we can fulfill (20 20) when in reality we can't